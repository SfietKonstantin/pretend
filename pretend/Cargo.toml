[package]
name = "pretend"
version = "0.1.0"
authors = ["Lucien XU <sfietkonstantin@free.fr>"]
edition = "2018"

[dependencies]
async-trait = { version = "0.1.48" }
http = "0.2"
pretend-macros = { path = "../pretend-macros" }
serde = "1.0"
thiserror = "1.0"
url = "2.2"

[dev-dependencies]
pretend-reqwest = { path = "../pretend-reqwest" }
pretend-testclient = { path = "../pretend-testclient" }
tokio = { version = "1.5", features = ["macros", "rt-multi-thread"] }
trybuild = "1.0"

