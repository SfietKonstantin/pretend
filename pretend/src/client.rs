//! Client traits
//!
//! This module contains traits to be implemented by a `pretend` client. `pretend` clients
//! implementations are responsible of executing the actual HTTP requests. Implementations
//! delegates to other crates.
//!
//! The central trait to implement is [`Client`]. This trait wraps the underlying
//! HTTP client and is responsible of executing the request via [`Client::execute`].
//! This method takes a method, url, header and body (as raw bytes) and should return
//! a response with raw bytes as body.
//!
//! Implementations should be marked with [`client::async_trait`] due to Rust limitations
//! with futures and traits.

pub use async_trait::async_trait;
pub use bytes::Bytes;
pub use http::Method;

use crate::{HeaderMap, Response, Result, Url};

/// `pretend` Client
///
/// See module level documentation for more information.
#[async_trait]
pub trait Client {
    /// Execute a request
    async fn execute(
        &self,
        method: Method,
        url: Url,
        headers: HeaderMap,
        body: Option<Bytes>,
    ) -> Result<Response<Bytes>>;
}
