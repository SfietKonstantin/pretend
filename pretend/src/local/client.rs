use super::Result;
use crate::{client, HeaderMap, Response, Url};
use async_trait::async_trait;
use bytes::Bytes;
use http::Method;

/// `pretend` local client
///
/// See module level documentation for more information.
#[async_trait(?Send)]
pub trait Client {
    /// Execute a request
    async fn execute(
        &self,
        method: Method,
        url: Url,
        headers: HeaderMap,
        body: Option<Bytes>,
    ) -> Result<Response<Bytes>>;
}

#[async_trait(?Send)]
impl<C> Client for C
where
    C: client::Client,
{
    async fn execute(
        &self,
        method: Method,
        url: Url,
        headers: HeaderMap,
        body: Option<Bytes>,
    ) -> Result<Response<Bytes>> {
        let response = C::execute(self, method, url, headers, body).await?;
        Ok(response)
    }
}
