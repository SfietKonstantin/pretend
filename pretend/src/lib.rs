//! `pretend` HTTP client
//!
//! # Examples
//!
//! ```rust
//! use pretend::{pretend, request, Pretend, Result, Url, UrlResolver};
//! # /*
//! use pretend_reqwest::Client;
//! # */
//! # use pretend_testclient::Client;
//! # use pretend::{HeaderMap, Response, StatusCode};
//!
//! #[pretend]
//! trait RestApi {
//!     #[request(method = "GET", path = "/get/{value}")]
//!     async fn api_get(&self, value: &str) -> Result<String>;
//! }
//!
//! #[tokio::main]
//! async fn main() {
//! # /*
//!     let client = Client::default();
//! # */
//! #    let client = Client::new(|_, _, _, _| {
//! #        let body = "Hello world";
//! #        let body = Some(body.as_bytes().to_vec());
//! #        Ok(Response::new(StatusCode::OK, HeaderMap::new(), body))
//! #    });
//!     let url = Url::parse("http://localhost").unwrap();
//!     let pretend = Pretend::for_client(client).with_url(url);
//!     let response = pretend.api_get("some-value").await.unwrap();
//!     assert_eq!(response, "Hello world");
//! }
//!
//! ```

// #[warn(missing_docs)]

pub mod client;
pub mod internal;

mod errors;

pub use self::errors::{Error, Result};
pub use http;
pub use http::{HeaderMap, StatusCode};
pub use pretend_codegen::{header, pretend, request};
pub use serde;
pub use serde::{Deserialize, Serialize};
pub use url;
pub use url::{ParseError, Url};

use crate::client::Client;
use serde::de::DeserializeOwned;
use std::result;

pub struct Response<T> {
    status: StatusCode,
    headers: HeaderMap,
    body: T,
}

impl<T> Response<T> {
    pub fn new(status: StatusCode, headers: HeaderMap, body: T) -> Self {
        Response {
            status,
            headers,
            body,
        }
    }

    pub fn status(&self) -> &StatusCode {
        &self.status
    }

    pub fn headers(&self) -> &HeaderMap {
        &self.headers
    }

    pub fn body(&self) -> &T {
        &self.body
    }

    pub fn into_body(self) -> T {
        self.body
    }

    pub fn map_body<F, U>(self, f: F) -> Response<U>
    where
        F: FnOnce(T) -> U,
    {
        Response {
            status: self.status,
            headers: self.headers,
            body: f(self.body),
        }
    }

    pub fn try_map_body<F, U>(self, f: F) -> Result<Response<U>>
    where
        F: FnOnce(T) -> Result<U>,
    {
        let body = f(self.body)?;
        Ok(Response {
            status: self.status,
            headers: self.headers,
            body,
        })
    }
}

pub trait ResolveUrl {
    fn resolve_url(&self, path: &str) -> result::Result<Url, ParseError>;
}

pub struct UrlResolver {
    base: Url,
}

impl UrlResolver {
    pub fn new(base: Url) -> Self {
        UrlResolver { base }
    }
}

impl ResolveUrl for UrlResolver {
    fn resolve_url(&self, path: &str) -> result::Result<Url, ParseError> {
        self.base.join(path)
    }
}

pub struct InvalidUrlResolver;

impl ResolveUrl for InvalidUrlResolver {
    fn resolve_url(&self, _: &str) -> result::Result<Url, ParseError> {
        Err(ParseError::EmptyHost)
    }
}

pub struct Pretend<C, R>
where
    C: Client + Send + Sync,
    R: ResolveUrl + Send + Sync,
{
    client: C,
    resolver: R,
}

impl<C, R> Pretend<C, R>
where
    C: Client + Send + Sync,
    R: ResolveUrl + Send + Sync,
{
    pub fn new(client: C, resolver: R) -> Pretend<C, R> {
        Pretend { client, resolver }
    }

    pub fn with_url(self, url: Url) -> Pretend<C, UrlResolver> {
        self.with_url_resolver(UrlResolver::new(url))
    }

    pub fn with_url_resolver<RR>(self, resolver: RR) -> Pretend<C, RR>
    where
        RR: ResolveUrl + Send + Sync,
    {
        Pretend {
            client: self.client,
            resolver,
        }
    }
}

impl<C> Pretend<C, InvalidUrlResolver>
where
    C: Client + Send + Sync,
{
    pub fn for_client(client: C) -> Pretend<C, InvalidUrlResolver> {
        Pretend {
            client,
            resolver: InvalidUrlResolver,
        }
    }
}

pub struct Json<T>
where
    T: DeserializeOwned,
{
    value: T,
}

impl<T> Json<T>
where
    T: DeserializeOwned,
{
    pub fn value(self) -> T {
        self.value
    }
}

pub enum JsonResult<T, E>
where
    T: DeserializeOwned,
    E: DeserializeOwned,
{
    Ok(T),
    Err(E),
}
