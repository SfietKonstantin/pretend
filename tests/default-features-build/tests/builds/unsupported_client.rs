use pretend::client::{BlockingClient, Bytes, Method};
use pretend::{Error, HeaderMap, Response, Result, Url};
use std::rc::Rc;
use thiserror::Error;

#[derive(Default, Debug, Error)]
#[error("Test error")]
struct TestError {
    data: Rc<i32>,
}

struct TestClient;

impl BlockingClient for TestClient {
    fn execute(
        &self,
        _: Method,
        _: Url,
        _: HeaderMap,
        _: Option<Bytes>,
    ) -> Result<Response<Bytes>> {
        Err(Error::response(TestError::default()))
    }
}

fn main() {}