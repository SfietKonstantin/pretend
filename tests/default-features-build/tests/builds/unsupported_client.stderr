error[E0277]: `Rc<i32>` cannot be sent between threads safely
  --> $DIR/unsupported_client.rs:22:29
   |
22 |         Err(Error::response(TestError::default()))
   |             --------------- ^^^^^^^^^^^^^^^^^^^^ `Rc<i32>` cannot be sent between threads safely
   |             |
   |             required by a bound introduced by this call
   |
   = help: within `TestError`, the trait `Send` is not implemented for `Rc<i32>`
note: required because it appears within the type `TestError`
  --> $DIR/unsupported_client.rs:8:8
   |
8  | struct TestError {
   |        ^^^^^^^^^
note: required by a bound in `pretend::Error::response`
  --> $DIR/errors.rs:73:30
   |
73 |     error::Error + 'static + Send + Sync
   |                              ^^^^ required by this bound in `pretend::Error::response`

error[E0277]: `Rc<i32>` cannot be shared between threads safely
  --> $DIR/unsupported_client.rs:22:29
   |
22 |         Err(Error::response(TestError::default()))
   |             --------------- ^^^^^^^^^^^^^^^^^^^^ `Rc<i32>` cannot be shared between threads safely
   |             |
   |             required by a bound introduced by this call
   |
   = help: within `TestError`, the trait `Sync` is not implemented for `Rc<i32>`
note: required because it appears within the type `TestError`
  --> $DIR/unsupported_client.rs:8:8
   |
8  | struct TestError {
   |        ^^^^^^^^^
note: required by a bound in `pretend::Error::response`
  --> $DIR/errors.rs:73:37
   |
73 |     error::Error + 'static + Send + Sync
   |                                     ^^^^ required by this bound in `pretend::Error::response`
